### DISCUSSION TITLE
EMQ X project demo sessions [old]

### DISCUSSION THREAD

**Discussion by zmstone** [emqx/emqx#4463] [Show and tell]
This thread recorded old demo sessions.
For new sessions since September 2021. see this thread instead: https://github.com/emqx/emqx/discussions/5622

EMQ X team would love to invite developers and users to join our online demo sessions.
In the online demo sessions, we will present some of the work we have been focusing on during the past few weeks. Comments and questions are welcome.

Demos will be scheduled [in this calendar](https://outlook.office365.com/calendar/published/be323b3a50ea4daeb04bf0c05ed94582@emqx.io/e4ef6b54cc7646ef86574323062cfdce15654452582260728863/calendar.html) 

Recordings are published to [YouTube](https://www.youtube.com/channel/UC5FjR77ErAxvZENEWzQaO5Q)

**Comment by zmstone** [comment]
# 20210406

## [@k32] Rlog (Async Mnesia DB replication) project updates

Rlog (short for Replicated Logs) is to provide an async replication mechanism for Mnesia tables in EMQ X, aiming to make EMQ X cluster more stable and scalable.

## [@qzhuyan] MQTT on QUIC

This is the third demo from the QUIC project, (flashback: the first was MsQuic NIF PoC, the second was the QUIC echo server from two weeks ago). We are now finally ready to demo MQTT on QUIC integrated into EMQ X.

## [@z8674558] HOCON, new config for EMQ X v5.0

[HOCON](https://github.com/lightbend/config) a JSON superset is now brought into the beam community.
Almost a drop-in replacement for the default cuttlefish parser used in EMQ X, we will start taking advantage of the HOCON features in EMQ X v5.0. 
We have worked on [emqx/hocon.git](https://github.com/emqx/hocon/) project for quite some time, however never presented it to the public. This is the time.

## [@terry-xiaoyu] jq.erl JQ for Erlang/Elixir

[jq](https://stedolan.github.io/jq/), de-facto JSON processing standard. Based on the open-source core Xinyu has built a NIF binding which will then be integrated into EMQ X's rule engine as a data processing function.

# Zoom Meeting

Topic: EMQ X sprint demoTime: Apr 6, 2021 09:00 Stockholm, 15:00 Beijing
Join Zoom Meeting https://zoom.us/j/91437642104?pwd=SVNSRmNEOWpHUm1jeFEwcHMrVTRQUT09
Meeting ID: 914 3764 2104  Passcode: 836954

**Comment by zmstone** [comment]
# 20210419

[Calendar, and zoom meeting link](https://outlook.office365.com/calendar/published/be323b3a50ea4daeb04bf0c05ed94582@emqx.io/e4ef6b54cc7646ef86574323062cfdce15654452582260728863/calendar.html)

Time: 09:00 GMT | 11:00 Stockholm | 17:00 Hangzhou | 18:00 Tokyo

## [@z8674558]: EMQ X config document auto generation

This is more of a technical presentation which may benefit EMQ X developers
and other users in the community if they happen to use [cuttlefish](https://github.com/emqx/cuttlefish)
for the current or next project.

When it comes to configuration, the 'source of truth' is usually the config file itself
and the code which parses the configs.
Maintaining the docs for configuration is quite a hassle if the doc is kept far from the 'source of truth',
doc-to-code discrepancy is the most common issue.

The feature which @z8674558 is going to demo, it will make use of the `@doc` annotations
in the config schema to generate [EMQ X configuration document](https://docs.emqx.io/en/broker/v4.3/configuration/configuration.html)

## [@k32]: Rlog (Async Mnesia db replication) project update (No.2)

Rlog (short for Replicated Logs) is to provide an async replication mechanism for Mnesia tables in EMQ X,
aiming to make EMQ X cluster more stable and scalable.

In the last demo @k32 presented some nice test results.
Now it's time to prove that EMQ X cluster will function well using rlog to replicate its database.

**Comment by zmstone** [comment]
# 2021-05-13 (Demo Day 003 - EN)

Time: 10:00 Stockholm, 16:00 Hangzhou/Beijing, 17:00 Tokyo
[Calendar](https://outlook.office365.com/calendar/published/be323b3a50ea4daeb04bf0c05ed94582@emqx.io/e4ef6b54cc7646ef86574323062cfdce15654452582260728863/calendar.html)
[Join Zoom Meeting](https://zoom.us/j/6013500474?pwd=T2pCYXArU1NuN2RTMGN3RFJYVnR5dz09) 

## Demos

### [@zmstone] Trie compaction: 
Released in v4.3.0. 6x times faster in a benchmark test for 5-level wildcard topic insert/delete performance. 10% slower in lookup performance

### [@qzhuyan]: Massive subscribe/unsubscribe performance improvements: 
Released in v4.2.12 e4.2.6 and v4.3.0. 10x times (maybe even more, re-running some tests) concurrent subscribe/unsubscribe request handling capacity

### [@k32] node-dump tool: 
One command line to collect runtime inforamtion and logs for issue reports

**Comment by tigercl** [comment]
## 2021-05-10 (Demo Day 001 - CN)
Time: 09:00 Stockholm, 15:00 Hangzhou/Beijing, 16:00 Tokyo
[Join Zoom Meeting](https://zoom.com.cn/j/98700719724)

#### [@Rory-Z] WebSocket connections use peer certificate as username
The WebSocket connection supports obtaining the peer certificate from the Proxy Protocol as the MQTT username.

#### [@wwhai]: Data Dump
Show how to dump data from a low version to our latest official version 4.3.0.

**Comment by zmstone** [comment]
# 2021-05-25  (EN 004)

Time: 10:00 Stockholm, 16:00 Hangzhou/Beijing, 17:00 Tokyo.
[Join Zoom Meeting](https://zoom.us/j/6013500474?pwd=T2pCYXArU1NuN2RTMGN3RFJYVnR5dz09)

[@zmstone] haproxy 2.4 sticky session towards EMQ X cluster
The very recently released haproxy 2.4 came with an exciting feature for MQTT community: it now can read MQTT packets for user-name, client ID etc. In this demo, we'll try to use client-id based sticky session for a consistent dispatch to EMQ X nodes in a cluster.

[@zmstone] send large file through EMQ X
v4.3.1 came with a performance fix which allowed sending large payload very quickly.
In this demo, we'll see how fast it is to send a 100+MB file through EMQ X using the newly introduced `--file` option in the `emqtt` client.

[@k32] RLOG storage optimization
Progress update on the RLOG project: previously the transaction log was stored in a regular `ram_copies` table that was cleaned using a special process. Given that the actual data in that table is not used, we optimized the it by introducing a "null" mnesia storage backend.

**Comment by tigercl** [comment]
## 2021-05-25 (Demo Day 002 - CN)
Time: 11:00 Stockholm, 17:00 Hangzhou/Beijing, 18:00 Tokyo
[Join Zoom Meeting](https://zoom.com.cn/j/95990227535)

### [@DDDHuang]: Support etcd3.4

Ekka cluster support etcd v3.4, reliable key & value distributed system. Demo shows service discover and auto cluster.

### [@DDDHuang]: Support StatsD

EMQ X system monitoring support stasd. Demo shows usage of stasd plugin and the view of metrics.

### [@terry-xiaoyu]: EMQ X Widget

The emqx_widget is an application that manages configuration specs and runtime states for components that need to be configured and manipulated from the emqx-dashboard.

It is intended to be used by resources, actions, acl, auth, backend_logics and more.

### [@tigercl]: New Authentication

Newly designed authentication feature for 5.0, this demo will show the basic usage of authentication chain and Mnesia authentication service.

### [@kerry-emqx]: New-looking of Modules in Dashboard

We entirely redesigned "modules" pages with simplicity, refreshing and consistency. Combined the `configuration` and the `management` to one page with multi-tabs that will never confuse anyone

**Comment by zmstone** [comment]
# 2021-06-08 (EN 005)
Time: 17:00(Tokyo); 10:00 (Stockholm); 16:00(Beijing/Hangzhou); 
[Zoom meeting](https://zoom.us/j/6013500474?pwd=T2pCYXArU1NuN2RTMGN3RFJYVnR5dz09)

@z8674558 HOCON Schema
EMQ X v5.0 will come with a unified validation for the two most important management interfaces: config file, and http API.
The library which helps us to achieve this goal is HOCON schema.

@k32 Async mnesia replication
This project is coming to the conclusion. The feature branch has been prepared for the public review.

@qzhuyan Quicer low level tracing and API updates
How to debug Quicer NIF internals and API updates

**Reply by zmstone** [reply]
We had to reschedule HOCON schema demo to the next sprint.

**Comment by tigercl** [comment]
## 2021-06-08(CN 003)

Time: 18:00(Tokyo); 11:00 (Stockholm); 17:00(Beijing/Hangzhou);
[Zoom meeting](https://zoom.com.cn/j/93884336040)

#### [@terry-xiaoyu]: emqx_data_bridge & emqx_resource。

The way resources are organized in 5.0 has changed. The "resources" part of the original rule engine is now in emqx_data_bridge and renamed to "data bridge" to manage the creation and maintenance of resources. A new behavior of emqx_resource has been added to the code level as a public lib to provide resource management capabilities for modules such as data bridging. This demo will demonstrate how to write a mysql bridge.

#### [@Rory-Z]: Demonstration of the basic rules of ACL chain

5.0 We have redesigned the ACL, now it supports more flexible rule configuration, adopts a more readable hocon format, supports updating through API, and supports querying customized ACL rules through external databases. This demo will demonstrate the operation of ACL rules and update ACL rules through API.

#### [@tigercl]: Authenticaiton HTTP API

Following the demo of the authentication function last time, we have made some minor adjustments to the authentication function and added the corresponding HTTP API. This demo will show how to use the authentication function through HTTP API and adjust the batch import of user information.

#### [@wivwiv]: UI design in EMQ X 5.0 Dashboard

Administrators can monitor cluster load conditions through EMQ X Dashboard, manage connected devices and configure various functions of EMQ X. The above operations also provide REST API to facilitate automated management, abandon complex command lines and configuration files, and provide real-time insights. Problems and bottlenecks in networked applications. This demo will share the UI and interaction design of some functions in EMQ X 5.0.

#### [@mengzhenyu]: GitHub actions execution process and various resource configuration explanations and demonstrations

Recently, we have added many automated integration tests to EMQ X to improve the stability of EMQ X. This demo will show how to use the yaml configuration file of GitHub actions workflows combined with jmeter scripts to implement automated integration tests.

**Comment by qzhuyan** [comment]
## 2021-07-06 (EN 006)

Time: 17:00(Tokyo); 10:00 (Stockholm); 16:00(Beijing/Hangzhou);
[Zoom meeting](https://us06web.zoom.us/j/83372515206)

@k32 RLOG project progress | mnesia:dirty_read function during shard bootstrap

**Comment by zmstone** [comment]
let's start a new thread and deprecate this one.
